File: 백준/Silver/2839. 설탕 배달/설탕 배달.cc
#include <iostream>
using namespace std;
constexpr int MAX = 5006;
constexpr int INF = 987654321;
int dp[MAX];
int n;
int f(int c) {
	if (c < 3 || c == 4 || c == 7) return INF;
	if (c == 3 || c == 5) return 1;
	if (dp[c] != INF) return dp[c];
	return dp[c] = min(f(c - 3) + 1, f(c - 5) + 1);
}
int main() {
	cin.tie(0); ios::sync_with_stdio(0);
	cin >> n;
	fill(dp, dp + MAX, INF);
	dp[n] = f(n);
	if (dp[n] == INF) dp[n] = -1;
	cout << dp[n];
}

File: 백준/Silver/2839. 설탕 배달/설탕 배달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static StringTokenizer st;
	static final int INF = 987654321;
	static final int RES = -1;
	static int dp[],n;
	static int f(int c) {
		if(c < 3 || c == 4 || c == 7) return INF;
		if(c == 3 || c == 5) return 1;
		if(dp[c] != INF) return dp[c];
		return dp[c] = Math.min(f(c-3)+1,f(c-5)+1);
	}
	public static void main(String[] args) throws NumberFormatException, IOException {
		n = Integer.parseInt(br.readLine());
		dp = new int[n+1];
		Arrays.fill(dp, INF);
		dp[n] = f(n);
		if(dp[n] == INF) dp[n] = -1;
		
		System.out.print(dp[n]);
	}
}